### PR TITLE
[C++ gRPC] Initial documentation of C++ gRPC

### PR DESCRIPTION
**Pull request by chadwalters**
Also, align usage of term 'stub' with gRPC convention and improve
consistency of C# docs with new C++ documentation.

### CODE REVIEW

File: doc/src/bond_over_grpc.md [hunk 1]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+
+    callback.wait();
+    ExampleResponse response = callback.response().Deserialize();
+    // Examine response here
+
+Note these APIs are significantly different from the APIs presented in the
+gRPC documentation; Bond-over-gRPC is attempting to provide a more
+straightforward API for aynchronous communication than gRPC currently presents
+in C++. The use of `wait_callback<T>` here is for illustrative purposes; any
+callback implementation with the same signature can be used. Note also the use
```
Comment (lalo) [inline doc/src/bond_over_grpc.md]: maybe we should mention that the callback signature should receive the input param type and grpc::Status returning void

`std::function<void(const ::bond::bonded< ::helloworld::HelloReply>&, const ::grpc::Status&)>`

File: doc/src/bond_over_grpc.md [hunk 2]
```diff
@@ -13,13 +13,13 @@ supercedes the (now deprecated) Bond Comm framework.
 The Bond IDL has been extended to support the definition of
 [services](compiler.html#service-definition) and
 [generic services](compiler.html#generic-service). These definitions are
-used by the Bond compiler to generate:
+used by the Bond compiler to generate classes that provide:
 
-* service stubs that can be used as the base for implementing services'
+* service base that can be used as the basis for implementing services'
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Copy edit: either "services base<em>s</em> that can be used as the basis..." or "_a_ service base..."

File: doc/src/bond_over_grpc.md [hunk 3]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+`gbc` will generate C++ classes for gRPC with the `--grpc` flag:
+
+    gbc c++ --grpc example.bond
+
+The key generated C++ classes for gRPC are:
+* A simple class with the name of the service (e.g.: `Example`). This class
+  provides some basic encapsulation of the server-side service base and the
+  client-side proxy stub.
+* The service base, which is an inner class named `Service` (e.g.:
+  `Example::Service`). [^1] This class has abstract methods
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Just so you're aware, this footnote syntax is Pandoc-specific. GitHub, for example, just has "\[^1\]" in the paragraph here and odd text at the bottom of the page.

(No change needed based on this comment, unless you want to rework the prose.)

File: doc/src/bond_over_grpc.md [hunk 4]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+            bond::ext::gRPC::unary_call<
+                bond::bonded<ExampleRequest>,
+                bond::bonded<ExampleResponse>> call) override
+        {
+            ExampleRequest request = call.request().Deserialize();
+            ExampleResponse reply;
+
+            // Service business logic goes here
+
+            call.Finish(bond::bonded<ExampleResponse >{response}, Status::OK);
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Minor: there an unneeded space in `bond::bonded<ExampleResponse >`

File: doc/src/bond_over_grpc.md [hunk 5]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+This service implementation is hooked up to a gRPC server as follows:
+
+    auto ioManager = std::make_shared<bond::ext::gRPC::io_manager>();
+    auto threadPool = std::make_shared<bond::ext::gRPC::thread_pool>();
+    const std::string server_address(Host + ":" + Port);
+
+    ExampleServiceImpl service;
+    bond::ext::gRPC::server_builder builder;
+    builder.SetThreadPool(threadPool);
+    const std::string server_address(Host + ":" + Port);
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Minor: two definitions of `server_address`.

File: doc/src/bond_over_grpc.md [hunk 6]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+in C++. The use of `wait_callback<T>` here is for illustrative purposes; any
+callback implementation with the same signature can be used. Note also the use
+of `bonded<T>` to wrap the request and response objects in several places;
+this allows for better control over the time of deserialization and also helps
+prevent slicing when using polymorphic Bond types. Finally, note that
+Bond-over-gRPC does not provide synchronous APIs in C++ by design.
+
+For more information about gRPC in C++, take a look at the
+[gRPC C++ tutorial](http://www.grpc.io/docs/tutorials/basic/c.html);
+however, keep in mind that the Bond-over-gRPC APIs diverge significantlty
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Copy edit: "significantly"

File: doc/src/bond_over_grpc.md [hunk 7]
```diff
@@ -90,21 +96,141 @@ On the client side, the proxy stub establishes a connection to the server like t
...
+- `examples/cpp/grpc/helloworld`
+
+
+[^1] Technically, `Service` is a type alias for
+`ServiceCore<bond::ext::gRPC::thread_pool>`. The `ServiceCore<T>` template
+can be used to customize the service implementation to use a different
+thread pool implementation.
+
+[^2] Likewise, `Client` is a type alias for
+`ClientCore<bond::ext::gRPC::thread_pool>`. The `ClientCore<>` template
```
Comment (chwarr) [inline doc/src/bond_over_grpc.md]: Minor: there an inconsistency between `ServiceCore<T>` above and just `ClientCore<>` here.

**Comment by lalo** [comment]
We could add a note on the difference between c# (exceptions) and c++ (status) failures.

File: doc/src/bond_over_grpc.md (hunk 1 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Let's add that in a later revision.

File: doc/src/bond_over_grpc.md (hunk 3 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Addressed.

File: doc/src/bond_over_grpc.md (hunk 4 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Addressed.

File: doc/src/bond_over_grpc.md (hunk 6 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Addressed.

File: doc/src/bond_over_grpc.md (hunk 5 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Addressed.

File: doc/src/bond_over_grpc.md (hunk 7 above)
Comment (chadwalters) [inline doc/src/bond_over_grpc.md]: Addressed.

**Comment by chwarr** [comment]
Merged into grpc++ in commit 8e1c32a3.

**Pull request closed without merging**